build(docs-infra): add deployment mode for release-candidate versions

gkalpak opened the pull request:
This PR adds the necessary build configuration/deployment mode (`rc`) for deploying release-candidate versions to angular.io. It is part of the work needed to prepare angular.io for [Versioning - Merging, Branching & Deployment](https://docs.google.com/document/d/197kVillDwx-RZtSVOBtPb4BBIAw0E9RT3q3v6DZkykU).

See individual commits for more info.

gkalpak commented:
This is blocked on renaming/creating the necessary Firebase projects.

mary-poppins commented:
You can preview e43030c at https://pr37280-e43030c.ngbuilds.io/.

mary-poppins commented:
You can preview 7bf2963 at https://pr37280-7bf2963.ngbuilds.io/.

petebacondarwin commented on aio/package.json:
```diff
@@ -175,4 +175,4 @@
   }
-}
\ No newline at end of file
+}
```
@mgol  would be pleased :-)

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -0,0 +1,151 @@
+        // Get the highest version.
+        .pop();
+
+  // Do not deploy as it is not the latest branch for the given major version.
```
```suggestion
  // Do not deploy if it is not the latest branch for the given major version.
```

gkalpak replied:
These comments were just moved over from `deploy-to-firebase.sh` and I didn't want to change them. This is removed in a subsequent commit, so we're good :smiley:

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -0,0 +1,151 @@
+}
+
+function yarn(cmd) {
+  return exec(`yarn ${cmd}`);
```
Does this output error messages to the console if the command fails?

gkalpak replied:
It does output a generic error message (something like `exec: internal error`), but not the actual error message.

Good catch! I'll update this.

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -0,0 +1,151 @@
+yarn('payload-size');
+
+// Deploy to Firebase.
+yarn(`firebase use "${projectId}" --token "${CI_SECRET_AIO_DEPLOY_FIREBASE_TOKEN}"`);
```
Does the `yarn()` helper display the secret token to the console?

gkalpak replied:
Without the `--silent/-s` flag, yarn does indeed echo the command (including expanded vars). So, it does indeed print the token to the output (even with out current setup). However, CircleCI masks secret env vars with `*****`. See https://circleci.com/gh/angular/angular/726807.

In this new implementation, we pass `silent: true` to ShellJS `exec()`, so the output (including the token) will not be printed to the output unless we explicitly `console.log()` the value returned by `yarn(...)`. (And even if we accidentally printed it, CircleCI would hopefully mask it for us.)

I can pass the `--silent` flag to make this even more safe.

gkalpak replied:
Actually, it seems that this is fixed in the latest version of ShellJS, so I'll update ShellJS from v0.7.7 to v0.8.4 instead.

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -89,15 +85,24 @@ if (CI_BRANCH === 'master') {
+      // This is the latest minor version for a major that is equal or greater than the stable major
+      // version, but not the stable version itself:
+      // Deploy as `rc`.
+      deployInfoPerTarget.rc;
```
👍  nice bit of logic here

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -7,150 +7,190 @@ const {cd, cp, exec: _exec, set} = require('shelljs');
-    `Deployment URL    : ${deployedUrl}\n`);
+function computeMajorVersion(branchName) {
+  return +branchName.split('.', 1)[0];
+}
```
OOC why move this above `computeInputVars()`?

gkalpak replied:
No idea. I usually order my function alphabetically, so I assume I originally named `computeInputVars()` something else that was lexicographically greater than `computeMajorVersion()` 🤷‍♂️

Moved below. Thx for catching 😇

petebacondarwin commented:
Oh and the first commit has a typo: `switch` -> `switches`.

gkalpak committed:
build(docs-infra): switch `deploy-to-firebase.sh` script to JS

This commit switches the `deploy-to-firebase.sh` script, that we use for
deploying angular.io to production, from Bash to JavaScript. This makes
the script easier to maintain.

For the same reasons, it also switches the `deploy-to-firebase.test.sh`
script, that we use for testing the `deploy-to-firebase` script, from
Bash to JavaScript (using jasmine as the test runner).

Finally, this commit also updates ShellJS to the latest version to get
better error messages (including the actual error) when `exec()` fails.

NOTE: Before switching the test script to JS, I verified that the new
      `deploy-to-firebase.js` script passed the tests with the old
      `deploy-to-firebase.test.sh` script.

gkalpak committed:
fixup! build(docs-infra): switch `deploy-to-firebase.sh` script to JS

gkalpak committed:
build(docs-infra): add new angular.io build configuration for release-candidates

This commit adds a new build configuration/deployment mode (`rc`) for
deploying release-candidate versions to angular.io.

This is part of the work needed to prepare angular.io for
[Versioning - Merging, Branching & Deployment](https://docs.google.com/document/d/197kVillDwx-RZtSVOBtPb4BBIAw0E9RT3q3v6DZkykU).

gkalpak committed:
build(docs-infra): add support for RC deployments to deployment script

This commit updates the angular.io deployment script
(`deploy-to-firebase.js`) to support deploying release-candidate
versions.

This is part of the work needed to prepare angular.io for
[Versioning - Merging, Branching & Deployment](https://docs.google.com/document/d/197kVillDwx-RZtSVOBtPb4BBIAw0E9RT3q3v6DZkykU).

gkalpak committed:
refactor(docs-infra): break up `deploy-to-firebase.js` script into functions

This commit breaks up the code in `deploy-to-firebase.js` script, that
we use for deploying angular.io to production, to smaller functions
(instead of a monolithic block). This makes the script easier to
maintain and also makes testing individual operations easier.

The commit also updates the `deploy-to-firebase.spec.js` spec file to
take advantage of the standalone functions to speed up testing by
calling the corresponding function instead of having to spawn a new
process and run the `deploy-to-firebase.js` script with the `--dry-run`
flag.

NOTE: Before updating the tests, I verified that the updated
      `deploy-to-firebase.js` script passed the old tests.

gkalpak committed:
fixup! refactor(docs-infra): break up `deploy-to-firebase.js` script into functions

gkalpak committed:
fixup! build(docs-infra): switch `deploy-to-firebase.sh` script to JS

gkalpak committed:
fixup! refactor(docs-infra): break up `deploy-to-firebase.js` script into functions

gkalpak commented:
Thx for the review, @petebacondarwin. I added two more commits addressing your feedback:
- 6a53ab96e17b5fc50a461e04fc79c26eba137b2f
- 40bf8c31323531cf3c2f4e88f5938c3e827d068b

PTAL

mary-poppins commented:
You can preview 40bf8c3 at https://pr37280-40bf8c3.ngbuilds.io/.

IgorMinar reviewed (approved):
This looks great!! But I don't think that we can land this now because rc-angular-io is not accessible to the aio-deploy@angular.io account for the reasons we discussed in the docs-infra meeting.

Should we still land this but change the project target for rc and next deployments to be the current  project? (I believe it's the aio-staging)

gkalpak commented:
It depends on what we want to do with RCs. The `aio-staging` project is indeed what will become `next-angular-io`, but currently we do not deploy RCs anywhere. If we want to keep it this way, then we could partially land this PR, but I don't think there is much benefit in doing so.

gkalpak commented:
Superceded by #39470.
